rust/src/quic/quic.rs

```diff
@@ -0,0 +1,336 @@
+            core::sc_app_layer_decoder_events_free_events(&mut self.events);
+        }
+        if let Some(state) = self.de_state {
+            core::sc_detect_engine_state_free(state);
+        }
+    }
+}
+
+impl Drop for QuicTransaction {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+
+pub struct QuicState {
+    tx_id: u64,
+    pub version: u32,
+    transactions: Vec<QuicTransaction>,
+}
+
+impl Default for QuicState {
+    fn default() -> Self {
+        Self {
+            tx_id: 0,
+            version: 0,
+            transactions: Vec::new(),
+        }
+    }
+}
+
+impl QuicState {
+    fn new() -> Self {
+        Self::default()
+    }
+
+    // Free a transaction by ID.
+    fn free_tx(&mut self, tx_id: u64) {
+        let tx = self
+            .transactions
+            .iter()
+            .position(|ref tx| tx.tx_id == tx_id + 1);
+        if let Some(idx) = tx {
+            let _ = self.transactions.remove(idx);
+        }
+    }
+
+    fn get_tx(&mut self, tx_id: u64) -> Option<&QuicTransaction> {
+        self.transactions.iter().find(|&tx| tx.tx_id == tx_id + 1)
+    }
+
+    fn new_tx(&mut self, header: QuicHeader, data: QuicData) -> QuicTransaction {
+        let mut tx = QuicTransaction::new(header, data);
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        return tx;
+    }
+
+    fn tx_iterator(
+        &mut self, min_tx_id: u64, state: &mut u64,
+    ) -> Option<(&QuicTransaction, u64, bool)> {
+        let mut index = *state as usize;
+        let len = self.transactions.len();
+
+        while index < len {
+            let tx = &self.transactions[index];
+            if tx.tx_id < min_tx_id + 1 {
+                index += 1;
+                continue;
+            }
+            *state = index as u64;
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
+        }
+
+        return None;
+    }
+
+    fn parse(&mut self, input: &[u8]) -> bool {
+        match QuicHeader::from_bytes(input, DEFAULT_DCID_LEN) {
+            Ok((rest, header)) => match QuicData::from_bytes(rest) {
+                Ok(data) => {
+                    self.version = header.version.into();
+
+                    let transaction = self.new_tx(header, data);
+                    self.transactions.push(transaction);
+
+                    return true;
+                }
+                Err(_e) => {
+                    return false;
+                }
+            },
+            Err(_e) => {
+                return false;
+            }
+        }
+    }
+}
+
+// C exports.
+export_tx_get_detect_state!(rs_quic_tx_get_detect_state, QuicTransaction);
+export_tx_set_detect_state!(rs_quic_tx_set_detect_state, QuicTransaction);
+
+#[no_mangle]
+pub extern "C" fn rs_quic_state_new(
+    _orig_state: *mut std::os::raw::c_void, _orig_proto: AppProto,
+) -> *mut std::os::raw::c_void {
+    let state = QuicState::new();
+    let boxed = Box::new(state);
+    return unsafe { transmute(boxed) };
```

Comment:
We've replaced transmute in this context with `Box::into_raw`, eg) https://github.com/OISF/suricata/blob/master/rust/src/dns/dns.rs#L646

Reply:
Maybe in the future we can look at blocking specific functions via: https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_method

Reply:
> Maybe in the future we can look at blocking specific functions via: https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_method

@jasonish should I let you do this clippy thing ?

Reply:
Good idea, and yeah.  Except I'm actually getting it to complain on any function I've tried.